File: Leetcode Daily Challenge/025-July-2025/011-meeting-rooms-iii.cpp
// https://leetcode.com/problems/meeting-rooms-iii/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    using ll = long long;
    using V  = vector<ll>;

    // TC : O(M*log(M) + M*log(N))
    // SC : O(M + N)
    int mostBooked(int n, vector<vector<int>>& meetings) {
        int m = meetings.size();

        sort(begin(meetings), end(meetings));

        priority_queue<int, vector<int>, greater<int>> freeRooms;
        // track the meeting which is currently happening : {meeEndingTime, roomNo.}
        priority_queue<V, vector<V>, greater<V>> ongoing;
        // track the number of times each room used
        vector<int> roomUsedTimesCnt(n, 0);

        for (int i = 0; i < n; i ++) freeRooms.push(i);

        for (auto &meet : meetings) {
            int currMeetStart = meet[0], currMeetEnd = meet[1];

            // Freeing up rooms whose meeting has ended
            while (!ongoing.empty() && ongoing.top()[0] <= currMeetStart) {
                freeRooms.push(ongoing.top()[1]);
                ongoing.pop();
            }

            if (!freeRooms.empty()) {
                int room = freeRooms.top();
                freeRooms.pop();
                roomUsedTimesCnt[room] ++;
                ongoing.push({currMeetEnd, room});

            } else {
                auto it = ongoing.top();
                ongoing.pop();

                ll earlyEndMeet = it[0];   // when room become free
                ll earlyEndRoom = it[1];   // No. of room which become free

                ll duration    = currMeetEnd - currMeetStart;
                ll nextMeetEnd = earlyEndMeet + duration;

                ongoing.push({nextMeetEnd, earlyEndRoom});
                roomUsedTimesCnt[earlyEndRoom] ++;
            }
        }

        int minRoomMaxUsed = 0, maxUsed = 0;
        for (int i = 0; i < n; i ++) {
            if (roomUsedTimesCnt[i] > maxUsed) {
                maxUsed = roomUsedTimesCnt[i];
                minRoomMaxUsed = i;
            }
        }
        return minRoomMaxUsed;
    }
};

File: Leetcode Daily Challenge/025-July-2025/011-meeting-rooms-iii.java
// https://leetcode.com/problems/meeting-rooms-iii/




import java.util.*;

class Solution {
    // TC : O(M*log(M) + M*log(N))
    // SC : O(M + N)
    public int mostBooked(int n, int[][] meetings) {

        Arrays.sort(meetings, (a, b) -> Integer.compare(a[0], b[0]));

        PriorityQueue<Integer> freeRooms = new PriorityQueue<>();

        PriorityQueue<long[]> ongoing = new PriorityQueue<>((a, b) -> {
            if (a[0] != b[0]) return Long.compare(a[0], b[0]);
            return Integer.compare((int)a[1], (int)b[1]);
        });

        int[] roomUsedTimesCnt = new int[n];

        for (int i = 0; i < n; i ++) {
            freeRooms.offer(i);
        }

        for (int[] meet : meetings) {
            int currMeetStart = meet[0], currMeetEnd = meet[1];

            // Freeing up rooms whose meeting has ended
            while (!ongoing.isEmpty() && ongoing.peek()[0] <= currMeetStart) {
                int room = (int)ongoing.poll()[1];
                freeRooms.offer(room);
            }

            if (!freeRooms.isEmpty()) {
                int room = freeRooms.poll();
                ongoing.offer(new long[]{currMeetEnd, room});
                roomUsedTimesCnt[room] ++;
            } else {
                long[] it = ongoing.poll();
                long earlyEnd  = it[0]; // next starting time of meeting
                int earlyRoom  = (int) it[1]; // available room for next meeting

                long duration   = currMeetEnd - currMeetStart;
                long newMeetEnd = earlyEnd + duration;

                roomUsedTimesCnt[earlyRoom] ++;
                ongoing.offer(new long[]{newMeetEnd, earlyRoom});
            }
        }

        int minRoom = 0, maxUsed = 0;
        for (int i = 0; i < n; i ++) {
            if (roomUsedTimesCnt[i] > maxUsed) {
                maxUsed = roomUsedTimesCnt[i];
                minRoom = i;
            }
        }
        return minRoom;
    }
}